### PR TITLE
Views management refactoring

### PR DESCRIPTION
**Pull request by ixrock**
Back to managing views with `webview` (iframe) in favour of smoothest clusters switching

### CODE REVIEW

File: src/main/index.ts [hunk 1]
```diff
@@ -41,8 +40,7 @@ async function main() {
   const updater = new AppUpdater()
   updater.start();
 
-  registerFileProtocol(appProto, app.getPath("userData"));
```
Comment (jakolehm) [inline src/main/index.ts:44]: Why `appProto` has been removed?

Comment (ixrock) [reply src/main/index.ts:44]: Because it's never used

Comment (jakolehm) [reply src/main/index.ts:44]: Might be a dumb question but how do we display uploaded cluster icons then?

Comment (Nokel81) [reply src/main/index.ts:44]: I switched to saving the base64 encoding of them in the user store

Comment (Nokel81) [reply src/main/index.ts:44]: It removes the filename issue that we were seeing with some unicode characters, and it removes the security problems when running on dev mode.

Comment (jakolehm) [reply src/main/index.ts:44]: > I switched to saving the base64 encoding of them in the user store

Ok, do we have a migration for that?

Comment (Nokel81) [reply src/main/index.ts:44]: I did not code one up, I should do that then

Comment (jakolehm) [reply src/main/index.ts:44]: > I did not code one up, I should do that then

Yes we should have a migration. Otherwise Lens will drop all the existing cluster icons on upgrade.

File: src/common/cluster-store.ts [hunk 2]
```diff
@@ -64,11 +62,13 @@ export class ClusterStore extends BaseStore<ClusterStoreModel> {
...
+      ipcRenderer.on("cluster:state", (event, state: ClusterState) => {
         this.applyWithoutSync(() => {
-          logger.debug(`[CLUSTER-STORE]: received state update for cluster=${clusterState.id}`, clusterState);
-          const cluster = this.getById(clusterState.id);
-          if (cluster) cluster.updateModel(clusterState)
+          logger.debug(`[CLUSTER-STORE]: received push-state at ${location.host}`, state);
+          const cluster = this.getById(state.id);
+          if (cluster) {
+            cluster.updateModel(state)
+          }
```
Comment (Nokel81) [inline src/common/cluster-store.ts]:
```suggestion
          this.getById(state.id)?.updatedModel(state);
```

File: src/common/ipc.ts [hunk 3]
```diff
@@ -2,94 +2,133 @@
...
 }
 
-export interface IpcMessageOpts<A extends any[] = any> {
+export interface IpcChannelResponse<T extends any[] = any, E = any> {
+  msgId: string;
+  data?: T;
+  error?: E;
+}
+
+export interface IpcChannelInit {
```
Comment (Nokel81) [inline src/common/ipc.ts]: This doesn't seem to be about the initial channel, wouldn't something like `IpcChannelOpts` be more fitting?

File: src/common/ipc.ts [hunk 4]
```diff
@@ -2,94 +2,133 @@
...
+  channel: IpcChannel; // main <-> renderer communication channel name
+  mode?: IpcMode; // default: "async", use "sync" as last resort: https://www.electronjs.org/docs/api/ipc-renderer#ipcrenderersendsyncchannel-args
+  handle?: (...args: any[]) => any; // main-process message handler
+  autoBind?: boolean; // auto-bind message handler in main-process, default: false
+  timeout?: number; // timeout for waiting response from the sender
+  once?: boolean; // todo: add support
+}
+
+export function createIpcChannel({ autoBind = true, mode = IpcMode.ASYNC, timeout = 0, handle, channel }: IpcChannelInit) {
+  channel = `${mode}:${channel}`
```
Comment (Nokel81) [inline src/common/ipc.ts]: Shouldn't it be an error to given `channel === ""`?

File: src/common/ipc.ts [hunk 5]
```diff
@@ -2,94 +2,133 @@
...
+          if (resolved) return;
+          res.msgId = req.msgId; // return back to sender to be able to handle response
+          resolved = true
+          logger.debug(`[IPC]: sending response to "${channel}"`, res);
+          if (mode === IpcMode.ASYNC) {
+            event.reply(channel, res);
+          }
+          if (mode === IpcMode.SYNC) {
+            event.returnValue = res;
+          }
```
Comment (Nokel81) [inline src/common/ipc.ts]: shouldn't this either be a `switch` or `if ... else if`?

File: src/common/ipc.ts [hunk 6]
```diff
@@ -2,94 +2,133 @@
...
+        msgId: getRandId({ prefix: "ipc-msg-id" }),
+        args: args,
+      }
+      if (mode === IpcMode.ASYNC) {
+        ipcRenderer.send(channel, req)
+      }
+      if (mode === IpcMode.SYNC) {
+        ipcRenderer.sendSync(channel, req)
+      }
+      return new Promise(async (resolve, reject) => {
```
Comment (Nokel81) [inline src/common/ipc.ts]: why are we returning a promise if the mode is `sync`?

File: src/common/user-store.ts [hunk 7]
```diff
@@ -71,11 +71,12 @@ export class UserStore extends BaseStore<UserStoreModel> {
     if (kubeConfig) {
       this.newContexts.clear();
       const localContexts = loadConfig(kubeConfig).getContexts();
-      console.log(localContexts)
-      localContexts
-        .filter(ctx => ctx.cluster)
-        .filter(ctx => !this.seenContexts.has(ctx.name))
-        .forEach(ctx => this.newContexts.add(ctx.name));
+      localContexts.forEach(({ cluster, name }) => {
```
Comment (Nokel81) [inline src/common/user-store.ts]: why this change, the older version is more clear

File: src/main/cluster.ts [hunk 8]
```diff
@@ -80,20 +84,15 @@ export class Cluster implements ClusterModel {
 
   @action
   async init(port: number) {
-    if (this.initialized) {
```
Comment (Nokel81) [inline src/main/cluster.ts:83]: Is it now save to init twice?

File: src/main/lens-proxy.ts [hunk 9]
```diff
@@ -117,26 +119,17 @@ export class LensProxy {
...
+    if (cluster) {
+      await cluster.contextHandler.ensureServer();
+      const proxyTarget = await this.getProxyTarget(req, cluster.contextHandler)
+      if (proxyTarget) {
+        // allow to fetch apis in "clusterId.localhost:port" from "localhost:port"
+        res.setHeader("Access-Control-Allow-Origin", this.origin);
+        return proxy.web(req, res, proxyTarget);
+      }
     }
+    this.router.route(cluster, req, res);
```
Comment (Nokel81) [inline src/main/lens-proxy.ts:132]: why route on an undefined cluster? Wouldn't it be better to change line 123 to be `if (!cluster) throw...` or `if (!cluster) return ...`?

File: src/main/routes/metrics-route.ts [hunk 10]
```diff
@@ -1,52 +1,34 @@
...
-
-    let metricsUrl: string
+    let prometheusPath: string
     let prometheusProvider: PrometheusProvider
     try {
-      const prometheusPath = await contextHandler.getPrometheusPath()
-      metricsUrl = `${kubeProxyUrl}/api/v1/namespaces/${prometheusPath}/proxy${cluster.getPrometheusApiPrefix()}/api/v1/query_range`
-      prometheusProvider = await contextHandler.getPrometheusProvider()
+      prometheusPath = await cluster.contextHandler.getPrometheusPath()
+      prometheusProvider = await cluster.contextHandler.getPrometheusProvider()
```
Comment (Nokel81) [inline src/main/routes/metrics-route.ts]: Shouldn't these be `await`-ed in together?

File: src/main/window-manager.ts [hunk 11]
```diff
@@ -1,68 +1,80 @@
...
-  protected activeView: BrowserWindow;
+  protected mainView: BrowserWindow;
   protected splashWindow: BrowserWindow;
-  protected noClustersWindow: BrowserWindow;
-  protected views = new Map<ClusterId, BrowserWindow>();
-  protected disposers: CallableFunction[] = [];
   protected windowState: windowStateKeeper.State;
 
-  constructor(protected proxyPort: number, showSplash = true) {
+  @observable activeClusterId: ClusterId;
```
Comment (Nokel81) [inline src/main/window-manager.ts:12]: If adding an observable field, shouldn't the class be marked as `observer`?

File: src/renderer/bootstrap.tsx [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    workspaceStore.load(),
+    clusterStore.load(),
+    i18nStore.init(),
+    themeStore.init(),
+  ]);
+
+  // init app's dependencies if any
+  if (App.init) {
+    await App.init();
+  }
```
Comment (Nokel81) [inline src/renderer/bootstrap.tsx:33]:
```suggestion
  await App.init?.call();
```

File: src/renderer/components/+apps-releases/release.store.ts [hunk 13]
```diff
@@ -58,7 +58,7 @@ export class ReleaseStore extends ItemStore<HelmRelease> {
     this.isLoading = true;
     let items;
     try {
-      const { isAdmin, allowedNamespaces } = getHostedCluster();
+      const { isAdmin, allowedNamespaces } = getHostedCluster()
```
Comment (Nokel81) [inline src/renderer/components/+apps-releases/release.store.ts:61]: why all the removing of semicolons?

File: src/renderer/components/cluster-manager/cluster-manager.tsx [hunk 14]
```diff
@@ -1,52 +1,116 @@
...
+  view: WebviewTag
+}
+
+const lensViews = observable.map<ClusterId, LensView>();
+
+export function hasLoadedView(clusterId: ClusterId): boolean {
+  return !!lensViews.get(clusterId)?.isLoaded;
+}
+
+// todo: figure out how to replace <webview>-tag to <iframe> with nodeIntegration=true
```
Comment (Nokel81) [inline src/renderer/components/cluster-manager/cluster-manager.tsx]: Maybe or https://github.com/react-native-community/react-native-webview

**Review by Nokel81 — COMMENTED** [review]
Looks good from a top level

File: src/common/ipc.ts (hunk 4 above)
Comment (ixrock) [reply src/common/ipc.ts:37]: doesn't really matter in this case, but ok

File: src/common/ipc.ts (hunk 3 above)
Comment (ixrock) [reply src/common/ipc.ts]: yep, sounds good

File: src/common/ipc.ts (hunk 4 above)
Comment (ixrock) [reply src/common/ipc.ts]: not critical imo, why not an empty string as a name?

File: src/common/ipc.ts (hunk 5 above)
Comment (ixrock) [reply src/common/ipc.ts]: doesn't really matter, only if it's a question of code-readability? :)

File: src/common/ipc.ts (hunk 6 above)
Comment (ixrock) [reply src/common/ipc.ts]: maybe we should, haven't even tested `sync` mode and it's better to use this "as last resort" as defined in docs:
https://www.electronjs.org/docs/api/ipc-renderer#ipcrenderersendsyncchannel-args

File: src/main/routes/metrics-route.ts (hunk 10 above)
Comment (ixrock) [reply src/main/routes/metrics-route.ts]: Not critical, but why not

File: src/main/window-manager.ts (hunk 11 above)
Comment (ixrock) [reply src/main/window-manager.ts:12]: No, `@observer` is used only with React-components. If component use in `render()` some mobx-observables only then it must have this decorator (to be able to re-render view).

File: src/renderer/bootstrap.tsx (hunk 12 above)
Comment (ixrock) [reply src/renderer/bootstrap.tsx:33]: Don't thinks it's a good idea imo..
`App.init?.call()` creates more cognitive load (since .call() rarely used in general or this way)

File: src/renderer/components/+apps-releases/release.store.ts (hunk 13 above)
Comment (ixrock) [reply src/renderer/components/+apps-releases/release.store.ts:61]: we don't have strict rules on that :) in general semicolon redundant and should be avoid imo..

File: src/renderer/components/cluster-manager/cluster-manager.tsx (hunk 14 above)
Comment (ixrock) [reply src/renderer/components/cluster-manager/cluster-manager.tsx]: it was more question "why <iframe> previously get to work with vue-app" 
p.s your link follows to `react-native-community`

**Comment by nevalla** [comment]
~~Cluster settings does not open when clicking from left menu. It's working if opening from system menu, but the layout is broken.~~ (FIXED)

File: src/common/ipc.ts (hunk 5 above)
Comment (nevalla) [reply src/common/ipc.ts]: IMHO `if ... else if` would be better here.

File: src/common/ipc.ts (hunk 6 above)
Comment (nevalla) [reply src/common/ipc.ts]: Just thinking that do we need `sync` mode here if we are only using async messaging?

**Comment by aleksfront** [comment]
> Cluster settings does not open when clicking from left menu. It's working if opening from system menu, but the layout is broken.

Layout is fixed.

File: src/main/cluster.ts (hunk 8 above)
Comment (ixrock) [reply src/main/cluster.ts:83]: No, init used only in single place in cluster-manager with that check

File: src/common/user-store.ts (hunk 7 above)
Comment (ixrock) [reply src/common/user-store.ts]: Not for me, sorry. Older version more like functional-style, but I would not say it's more readable (for humans) + there is no need to run multiple cycles on what's can be updated in one run.

File: src/common/ipc.ts (hunk 6 above)
Comment (ixrock) [reply src/common/ipc.ts]: Well, this `IPC` helper is intended to be multi-functional, so I would keep it, but have to test if it's really works as expected.

File: src/main/window-manager.ts (hunk 11 above)
Comment (Nokel81) [reply src/main/window-manager.ts:12]: Ah thanks for the information, I did not know that

File: src/common/user-store.ts (hunk 7 above)
Comment (jakolehm) [reply src/common/user-store.ts]: I think this is matter of taste. I would revert this change as it's not related to views management at all, right?

File: src/main/cluster-manager.ts [hunk 15]
```diff
@@ -50,23 +42,8 @@ export class ClusterManager {
...
-      }
-    } else {
-      const id = req.headers.host.split(".")[0]
-      cluster = this.getCluster(id)
-    }
-
-    return cluster;
+    logger.debug(`getClusterForRequest(): ${req.headers.host}${req.url}`)
+    const clusterId = req.headers.host.split(".")[0]
+    return this.getCluster(clusterId)
```
Comment (jakolehm) [inline src/main/cluster-manager.ts]: Why logic has been removed? Don't we use `127.0.0.1:<port>/uuid` anymore anywhere?

Comment (ixrock) [reply src/main/cluster-manager.ts]: Seems so.. Looks like outdated code from very beginning of lens-app.

Comment (jakolehm) [reply src/main/cluster-manager.ts]: I asked because I `port/uuid` addresses were used when doing requests outside of renderer (wildcard dns to localhost works only within browser scope). Anyway I don't like to see changes that are totally outside of PR scope, it makes review unnecessary complicated.

Comment (ixrock) [reply src/main/cluster-manager.ts]: Ok, reverted back. Someone should check if this is actually works, cause top comment:
`// lens-server is connecting to 127.0.0.1:<port>/<uid>`
which unclear in terms of what you saying how this might be used outside.

File: src/common/ipc.ts (hunk 4 above)
Comment (Nokel81) [reply src/common/ipc.ts]: Just seems like a weird that a channel called `"sync:"` or `"async:"` could be made. Another possible solution could be 

```javascript
channel = `${mode}:${channel || "default"}`
```

File: src/renderer/bootstrap.tsx (hunk 12 above)
Comment (Nokel81) [reply src/renderer/bootstrap.tsx:33]: Makes sense

**Review by nevalla — APPROVED** [review]
